src/main/java/io/cdap/plugin/batch/joiner/Lookup.java

```diff
@@ -21,133 +21,272 @@
+                                                              Collections.singletonList(stagedInputField),
+                                                              outputFieldName);
+        operations.add(identity);
+        continue;
+      }
+
+      String operationName = String.format("Rename %s", stagedInputField);
+
+      FieldOperation transform = new FieldTransformOperation(operationName, RENAME_OPERATION_DESCRIPTION,
+                                                             Collections.singletonList(stagedInputField),
+                                                             outputFieldName);
+      operations.add(transform);
     }
-    StructuredRecord.Builder builder = StructuredRecord
-      .builder(Schema.recordOf("record", Schema.Field.of("id", fieldSchema)));
-    // depending on stage get the join key value
-    Object value = isLookupStage ? lookupValue : structuredRecord.get(config.getInputKeyField());
-    builder.set("id", value);
-    return Collections.singletonList(builder.build());
-  }
 
-  @Override
-  public JoinConfig getJoinConfig() throws Exception {
-    return joinConfig;
+    return operations;
   }
 
+  @Nullable
   @Override
-  public StructuredRecord merge(StructuredRecord joinKey, Iterable<JoinElement<StructuredRecord>> iterable)
-    throws Exception {
-    StructuredRecord.Builder builder = StructuredRecord.builder(Schema.parseJson(config.schema));
-    builder.set(config.getOutputField(), config.getDefaultValue());
-    final Iterator<JoinElement<StructuredRecord>> iterator = iterable.iterator();
-    while (iterator.hasNext()) {
-      final JoinElement<StructuredRecord> next = iterator.next();
-      if (next.getStageName().equals(config.getLookupDataset())) {
-        builder.set(config.getOutputField(), next.getInputRecord().get(config.getLookupValueField()));
+  public JoinDefinition define(AutoJoinerContext context) {
+    FailureCollector collector = context.getFailureCollector();
+    boolean hasUnknownInputSchema = context.getInputStages().values().stream().map(JoinStage::getSchema)
+      .anyMatch(Objects::isNull);
+    if (hasUnknownInputSchema) {
+      collector.addFailure("Schemas for all input stages must be provided.", null)
+        .withConfigProperty(Config.OUTPUT_SCHEMA);
+      return null;
+    }
+
+    if (config.fieldsContainMacros()) {
+      return null;
+    }
+
+    Set<String> requiredStages = context.getInputStages().keySet().stream()
+      .filter(stageName -> !stageName.equals(config.getLookupDataset())).collect(Collectors.toSet());
+    Set<String> broadcastStages = Collections.singleton(config.getLookupDataset());
+    List<JoinStage> inputs = new ArrayList<>(context.getInputStages().size());
+    boolean useOutputSchema = false;
+    for (JoinStage joinStage : context.getInputStages().values()) {
+      inputs.add(JoinStage.builder(joinStage)
+                   .setRequired(requiredStages.contains(joinStage.getStageName()))
```

Comment:
Isn't there exactly 1 broadcast stage and 1 required stage? If yes, there is no need for a set - just a string comparison will be sufficient.

Reply:
updated